analyzers/src/SonarAnalyzer.CSharp/Rules/TimestampsShouldNotBeUsedAsPrimaryKeys.cs

```diff
@@ -0,0 +1,69 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class TimestampsShouldNotBeUsedAsPrimaryKeys : SonarDiagnosticAnalyzer
+{
+    private const string DiagnosticId = "S3363";
+    private const string MessageFormat = "Timestamps should not be used as primary keys";
+
+    private static readonly KnownType[] TemporalTypes = new[]
+    {
+        KnownType.System_DateTime,
+        KnownType.System_DateTimeOffset,
+        KnownType.System_TimeSpan
+    };
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(c =>
```

Comment:
_Open question/educational_: what if we could implement a cost-effective check on the using statements in order to determine if Entity Framework or Entity Framework Core dependencies are present? Although I'm not sure about the complexity of retrieving global statements, this check could be triggered just once per file, even before diving further into analyzing classes, properties, and attributes.

Reply:
An entity class doesn't necessarily need to be marked with attributes in Entity Framework. We can have this entity in one file without any using directives from EF:

```
public class Person
{
   public int Id { get; set; }
   public string FirstName { get; set; }
   public string LastName { get; set; }
   public DateOnly DateOfBirth { get; set; }
}
```

And in another file:

```
using Microsoft.EntityFrameworkCore;

public class DatabaseContext : DbContext
{
    public DbSet<Person> People { get; set; }
    ...
}
```

I'm going to opt for [the suggestion](https://github.com/SonarSource/sonar-dotnet/pull/7471#discussion_r1235783119) from @csaba-sagi-sonarsource to remove all access to the semantic model for this rule to optimize performance and document it.

Reply:
Another thing here that can trick us is the usage of global using. In that case the using Microsoft.EntityFrameworkCore can be in a completely different file too.